carbon-p2-plugin/src/main/java/org/wso2/maven/p2/utils/P2Utils.java

```diff
@@ -0,0 +1,188 @@
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.transform.OutputKeys;
+import javax.xml.transform.Transformer;
+import javax.xml.transform.TransformerConfigurationException;
+import javax.xml.transform.TransformerException;
+import javax.xml.transform.TransformerFactory;
+import javax.xml.transform.dom.DOMSource;
+import javax.xml.transform.stream.StreamResult;
+
+
+/**
+ * The class contains operations related to Categories, which is out of the scope for current set of tasks. Thus
+ * the refactoring is done for a minimum for the moment.
+ */
+public class P2Utils {
+    private static String[] matchList = new String[]{"perfect", "equivalent", "compatible", "greaterOrEqual", "patch",
+            "optional"};
+    private static final String DEFAULT_ENCODING = "UTF-8";
+
+    public static int getLastIndexOfProperties(File p2InfFile) throws IOException {
+        int min = -1;
+        if (p2InfFile.exists()) {
+
+            try (InputStreamReader reader = new InputStreamReader(new FileInputStream(p2InfFile), DEFAULT_ENCODING);
+                 BufferedReader in = new BufferedReader(reader)) {
+                String line;
+                while ((line = in.readLine()) != null) {
+                    String[] split = line.split("=");
+                    String[] split2 = split[0].split(Pattern.quote("."));
+                    if (split2[0].equalsIgnoreCase("properties")) {
+                        int index = Integer.parseInt(split2[1]);
+                        if (index > min) {
+                            min = index;
+                        }
+                    }
+                }
+            }
+        }
+        return min;
+    }
+
+    public static boolean isMatchString(String matchStr) {
+        for (String match : matchList) {
+            if (matchStr.equalsIgnoreCase(match)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    public static String getMatchRule(String matchStr) {
+        if (isPatch(matchStr)) {
+            return "perfect";
+        }
+        for (String match : matchList) {
+            if (matchStr.equalsIgnoreCase(match)) {
+                return match;
+            }
+        }
+        return null;
+    }
+
+    public static boolean isPatch(String matchStr) {
+        return matchStr.equalsIgnoreCase("patch");
+    }
+
+    public static void createCategoryFile(MavenProject project, List categories, File categoryFile)
+            throws ParserConfigurationException, TransformerException, MojoExecutionException {
+
+        Map featureCategories = new HashMap();
```

Comment:
During a technical discussion had with Sameera it was decided just to move and not to touch code related to categories. Thus what you see is non-refactored only migrated code.